Add documentation indicating how to create and execute tests

hugo-dc opened the pull request.

hugo-dc committed:
Add documentation indicating how to create and execute tests

jakelang commented on src/GeneralStateTestsFiller/stEWASMTests/README.md:
```diff
@@ -0,0 +1,317 @@
+## eWasm Tests
```
it is "ewasm" now :)

axic commented on src/GeneralStateTestsFiller/stEWASMTests/README.md:
````diff
@@ -0,0 +1,317 @@
+  )
+```
+
+Code is wrapped in a `module`
````
This probably should say something like "testeth detects a source starting with `(module` as wasm and will automatically compile it"

axic commented on src/GeneralStateTestsFiller/stEWASMTests/README.md:
```diff
@@ -0,0 +1,317 @@
+- secretKey - using
+  `45a915e4d060149eb4365960e6a7a45f334393093061116b197e3240065ff2d8` which is
+  the secret key for account
+  `45a915e4d060149eb4365960e6a7a45f334393093061116b197e3240065ff2d8`
```
This seems to be a private key not an address?

hugo-dc replied:
good catch

zixuanzh commented:
should also run `git submodule update --init` inside `cpp-ethereum/hera`

hugo-dc commented:
@zixuanzh that's correct, I'll fix it.

hugo-dc committed:
change title as ewasm, clarify module, add correct account for private key, specify we need to update hera submodules too

hugo-dc commented:
@lrettig if there are no more comments about this document it would be good to merge it.

lrettig commented:
Sure thing @hugo-dc, let me take one final look at this. Thanks again for writing it!

lrettig commented on src/GeneralStateTestsFiller/stEWASMTests/README.md:
```diff
@@ -0,0 +1,318 @@
+[9]: https://github.com/ewasm/tests/blob/wasm-tests/src/GeneralStateTestsFiller/stEWASMTests/callFiller.yml#L1
+[10]: https://github.com/ewasm/tests/blob/wasm-tests/src/GeneralStateTestsFiller/stEWASMTests/callFiller.yml#L88
+[11]: https://github.com/ewasm/design/blob/master/eth_interface.md
+[12]: https://github.com/ewasm/hera/blob/master/circle.yml#L65
```
Some links are broken, including this one to `circle.yml` since the file has changed. Could you replace this with a permalink e.g. (in this case) https://github.com/ewasm/hera/blob/5eb92df16b49b673852ad8b6d4b0f469a13d3a23/circle.yml#L70 ?

lrettig commented on src/GeneralStateTestsFiller/stEWASMTests/README.md:
```diff
@@ -0,0 +1,318 @@
+### 2. pre - state before the transaction (pre-state)
+
+Here we define accounts and the initial values for each one, for almos all test
+cases we are using the account `a94f5374fce5edbc8e2a8697c15331677e6ebf0b` this
```
Could you mention that there is nothing special about this account, and that we just use the same account in all of the tests by convention?

lrettig commented on src/GeneralStateTestsFiller/stEWASMTests/README.md:
```diff
@@ -0,0 +1,318 @@
+[7]: https://github.com/ewasm/tests/blob/wasm-tests/GeneralStateTests/stEWASMTests/call.json#L21
+[8]: https://github.com/ewasm/tests/tree/wasm-tests/src/GeneralStateTestsFiller/stEWASMTests
+[9]: https://github.com/ewasm/tests/blob/wasm-tests/src/GeneralStateTestsFiller/stEWASMTests/callFiller.yml#L1
+[10]: https://github.com/ewasm/tests/blob/wasm-tests/src/GeneralStateTestsFiller/stEWASMTests/callFiller.yml#L88
```
Broken link

lrettig commented on src/GeneralStateTestsFiller/stEWASMTests/README.md:
```diff
@@ -0,0 +1,318 @@
+
+In [GeneralStateTests][3] you can find the test cases specific for
+eWasm: [stEWASMTests][4], this directory contains the "filled" json files, those
+files are read by testeth, testeth reads the [pre-state][5],
```
Rather than linking to external files, could we include sample pre-state, transaction, and expect sections inline here as you've done below?

lrettig commented on src/GeneralStateTestsFiller/stEWASMTests/README.md:
```diff
@@ -0,0 +1,318 @@
+
+#### code
+
+Example of wast code: 
```
Could you include a link to the WASM spec or docs that explains the structure of a WASM module in more detail? Including imports, memory, data, exports, func, etc.

lrettig commented on src/GeneralStateTestsFiller/stEWASMTests/README.md:
````diff
@@ -0,0 +1,318 @@
+  )
+```
+
+`testeth` detects source starting with `(module` as wasm and will automatically compile it.
````
How is it compiled?

hugo-dc replied:
@lrettig not sure if I understand it correctly, want me to specify what testeth uses to compile wat to wasm bytecode?

lrettig replied:
I just want to understand how this happens. Which compiler does it pass through? Maybe it's not critical for this doc

lrettig commented on src/GeneralStateTestsFiller/stEWASMTests/README.md:
````diff
@@ -0,0 +1,318 @@
+      balance: "100000000001"
+```
+
+For the expect section values of indexes and network remain the same in all test
````
Good to point out that for ewasm we only ever want/need to pass >= Byzantium tests.

lrettig commented on src/GeneralStateTestsFiller/stEWASMTests/README.md:
```diff
@@ -0,0 +1,318 @@
+
+## Filling and executing the tests
+
+Filling the test is the process to take the source Filler files and generate
```
Please point out here that, if you change a test or write a new one, your git commit must include an updated, filled test file as well. Also might be good to link to the cpp-ethereum/testeth docs here, I think there's a doc explaining what test filling is.

lrettig commented on src/GeneralStateTestsFiller/stEWASMTests/README.md:
````diff
@@ -0,0 +1,318 @@
+git submodule update --init
+```
+
+- update hera to the latest version (or the specific hera version you want to
````
This is confusing. Updating hera like this shouldn't be necessary if the user has already run `git submodule update --init` (and maybe with `--recursive` as well--actually I'm not sure this is necessary for hera). Can we just keep the right version of hera pinned at all times? Maybe the note should say, "If you want to test against a specific version of Hera, you can do it like this..."

lrettig commented on src/GeneralStateTestsFiller/stEWASMTests/README.md:
````diff
@@ -0,0 +1,318 @@
+```
+mkdir build
+cd build
+cmake -DHERA=ON -DHERA_DEBUGGING=ON .. && make -j4
````
I think we should recommend building Hera as a dynamic module by default, since it's much easier then to rebuild only hera if it changes.

lrettig commented on src/GeneralStateTestsFiller/stEWASMTests/README.md:
````diff
@@ -0,0 +1,318 @@
+You can fill the test using the following command:
+
+```
+ETHEREUM_TEST_PATH=/ewasm/tests ./testeth -t GeneralStateTests/stEWASMTests -- --filltests --vm hera --singlenet "Byzantium" --singletest mytest
````
The `--jsontrace ''` flag is also sometimes useful, might be worth documenting it here.

lrettig reviewed (changes requested):
Added comments. Thanks @hugo-dc this is an awesome start to the documentation!

hugo-dc committed:
Change links to files to permalinks

hugo-dc committed:
- Fix broken links, use permalinks instead
- Add link to the WASM spec
- Point out that the filled file has to be added to each commit when changing or creating a new test case
- Add a link to the testeth documentation indicating how to fill tests
- Clarify you can build diferent versions of hera
- Indicate how to build hera as a shared library

hugo-dc committed:
specify that any account can be used for testing, we are executing tests for byzantium, jsontrace, add filled sections directly instead of linking external file

hugo-dc commented:
@lrettig I made some changes according to your comments, please take a look. Thank You!

Commit:
Update README.md

st/eWASM/ewasm and tweak one sentence.

lrettig commented:
Made one small change. Otherwise LGTM. Thanks for the hard work on this doc Hugo!

hugo-dc commented:
@lrettig Thanks for reviewing!
